From dc485db689d532b6a616b07b48baacacd07284ad Mon Sep 17 00:00:00 2001
From: Samuel Dionne-Riel <samuel@dionne-riel.com>
Date: Sun, 14 Apr 2019 20:09:41 -0400
Subject: [PATCH 2/2] !!! Loads FDT from qualcomm-specific format of android
 bootimg

---
 common/image-fdt.c | 43 +++++++++++++++++++++++++++++--------------
 1 file changed, 29 insertions(+), 14 deletions(-)

diff --git a/common/image-fdt.c b/common/image-fdt.c
index 01186aeac7..378f9a285b 100644
--- a/common/image-fdt.c
+++ b/common/image-fdt.c
@@ -458,21 +458,36 @@ int boot_get_fdt(int flag, int argc, char * const argv[], uint8_t arch,
 			goto no_fdt;
 		}
 #ifdef CONFIG_ANDROID_BOOT_IMAGE
-	} else if (genimg_get_format(buf) == IMAGE_FORMAT_ANDROID) {
-		struct andr_img_hdr *hdr = buf;
-		ulong fdt_data, fdt_len;
-
-		if (android_image_get_second(hdr, &fdt_data, &fdt_len) != 0)
-			goto no_fdt;
+// FIXME : use `android_image_get_dt` when this fails
+//	} else if (genimg_get_format(buf) == IMAGE_FORMAT_ANDROID) {
+//		struct andr_img_hdr *hdr = buf;
+//		ulong fdt_data, fdt_len;
+//
+//		if (android_image_get_second(hdr, &fdt_data, &fdt_len) != 0)
+//			goto no_fdt;
+//
+//		fdt_blob = (char *)fdt_data;
+//		if (fdt_check_header(fdt_blob) != 0)
+//			goto no_fdt;
+//
+//		if (fdt_totalsize(fdt_blob) != fdt_len)
+//			goto error;
+//
+//		debug("## Using FDT found in Android image second area\n");
+	} else if (/* FIXME: detect presence of dt? */ 1) {
+			ulong fdt_len;
+			android_image_get_dt((void *)images->os.start,
+				&fdt_addr, &fdt_len);
+			fdt_blob = (char *)fdt_addr;
+
+			debug("*  fdt: raw FDT blob\n");
+			printf("## Flattened Device Tree blob at %08lx\n",
+				   (long)fdt_addr);
 
-		fdt_blob = (char *)fdt_data;
-		if (fdt_check_header(fdt_blob) != 0)
-			goto no_fdt;
-
-		if (fdt_totalsize(fdt_blob) != fdt_len)
-			goto error;
-
-		debug("## Using FDT found in Android image second area\n");
+			if (fdt_totalsize(fdt_blob) != fdt_len) {
+				fdt_error("fdt size != image size");
+				goto error;
+			}
 #endif
 	} else {
 		debug("## No Flattened Device Tree\n");
-- 
2.18.1

