From 88786b4a40ffa7bc75598252856e9640bcd1b27b Mon Sep 17 00:00:00 2001
From: Samuel Dionne-Riel <samuel@dionne-riel.com>
Date: Sun, 28 Jun 2020 21:05:41 -0400
Subject: [PATCH] mobile-nixos: Add identifier nodes to root node

---
 arch/arm64/boot/dts/E262L.dts | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/arch/arm64/boot/dts/E262L.dts b/arch/arm64/boot/dts/E262L.dts
index e2d257ba8847..6fad24cb99a2 100755
--- a/arch/arm64/boot/dts/E262L.dts
+++ b/arch/arm64/boot/dts/E262L.dts
@@ -1520,3 +1520,9 @@
 	cd_level = /bits/ 8 <MSDC_CD_HIGH>;  
 };
 /* lihaiyan@wind-mobi.com  for SD card end 20170207 */
+
+/* Mobile NixOS properties */
+/ {
+	mobile-nixos,device-name = "asus-x018d";
+	mobile-nixos,display-name = "Asus Zenfone Max Plus (M1)";
+};
-- 
2.25.4

