From 33797e8f63aa72659af22f56e888c63d98b15eb1 Mon Sep 17 00:00:00 2001
From: Samuel Dionne-Riel <samuel@dionne-riel.com>
Date: Mon, 29 Jun 2020 01:02:46 -0400
Subject: [PATCH 2/2] E262L: Green LED now defaults to on

---
 arch/arm64/boot/dts/E262L.dts | 1 +
 1 file changed, 1 insertion(+)

diff --git a/arch/arm64/boot/dts/E262L.dts b/arch/arm64/boot/dts/E262L.dts
index e2d257ba8847..228cab8bd715 100755
--- a/arch/arm64/boot/dts/E262L.dts
+++ b/arch/arm64/boot/dts/E262L.dts
@@ -43,6 +43,7 @@
 		led_mode = <3>;
 		data = <2>;
 		pwm_config = <0 0 0 0 0>;
+		linux,default-trigger = "default-on";
 	};
 //ranyanhao@wind-mobi.com 20160221 end
 	
-- 
2.25.4

