diff --git a/drivers/staging/rtl8723cs-new/Makefile b/drivers/staging/rtl8723cs-new/Makefile
index 17e66758e6fb..626119f460be 100644
--- a/drivers/staging/rtl8723cs-new/Makefile
+++ b/drivers/staging/rtl8723cs-new/Makefile
@@ -48,7 +48,7 @@ CONFIG_SDIO_HCI = y
 CONFIG_GSPI_HCI = n
 ########################## Features ###########################
 CONFIG_MP_INCLUDED = y
-CONFIG_POWER_SAVING = y
+CONFIG_POWER_SAVING = n
 CONFIG_USB_AUTOSUSPEND = n
 CONFIG_HW_PWRP_DETECTION = n
 CONFIG_WIFI_TEST = n
