diff --git a/arch/arm64/boot/dts/rockchip/rk3399-pinephone-pro.dts b/arch/arm64/boot/dts/rockchip/rk3399-pinephone-pro.dts
index 520d998bae9a..2234b12dd546 100644
--- a/arch/arm64/boot/dts/rockchip/rk3399-pinephone-pro.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3399-pinephone-pro.dts
@@ -750,14 +750,13 @@ regulator-state-mem {
  * VCC2V8_DVP   af    - autoenabled by VCC1V8_DVP   depends on VCC3V3_SYS
  * AVDD2V8_DVP        - autoenabled by VCC1V8_DVP   depends on VCC3V3_SYS
  *
- * mipi0  = imx258
- *   reset = gpio1 A0
+ * isp1 = imx258
+ *   reset = GPIO1 A0
  *   powerdown = GPIO2_B4
  *
- * mipi1  = OV5640
+ * isp0 = OV8858
  *   powerdown = GPIO2_D4
- *
- *   bt656-supply = <&vcc1v8_dvp>;
+ *   reset = GPIO1_A4
  */
 &i2c1 {
 	status = "okay";
@@ -783,7 +782,7 @@ wcam: camera@1a {
 		clock-names = "xvclk";
 
 		vif-supply = <&vcc1v8_dvp>;
-		/*XXX: also depends on vcca1v8_codec for I2C bus power */
+		/*XXX: also depends on vcca1v8_codec for I2C bus power (currently always on) */
 
 		reset-gpios = <&gpio1 RK_PA0 GPIO_ACTIVE_LOW>;
 		powerdown-gpios = <&gpio2 RK_PD4 GPIO_ACTIVE_HIGH>;
@@ -793,7 +792,7 @@ port {
 			wcam_out: endpoint {
 				remote-endpoint = <&mipi_in_wcam>;
 				data-lanes = <1 2 3 4>;
-				link-frequencies = /bits/ 64 <320000000>;
+				//link-frequencies = /bits/ 64 <320000000>;
 				//link-frequencies = /bits/ 64 <633600000>;
 			};
 		};
@@ -969,7 +968,7 @@ &io_domains {
 	gpio1830-supply = <&vcc_3v0>;
 };
 
-&isp0 {
+&isp1 {
 	status = "okay";
 
 	ports {
@@ -984,11 +983,11 @@ mipi_in_wcam: endpoint@0 {
 };
 
 &mipi_dphy_rx0 {
-	status = "okay";
+	//status = "okay";
 };
 
 &isp0_mmu {
-	status = "okay";
+	//status = "okay";
 };
 
 &mipi_dsi {
