From 6eb5ae25558b2e9bb79158ca39a52071dceabbe0 Mon Sep 17 00:00:00 2001
From: Samuel Dionne-Riel <samuel@dionne-riel.com>
Date: Tue, 31 Aug 2021 02:00:51 -0400
Subject: [PATCH] mm/shmem: Fix incomplete backport with TMPFS_POSIX_ACL

---
 mm/shmem.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mm/shmem.c b/mm/shmem.c
index 8e0b54348275f..b70bca22719b9 100644
--- a/mm/shmem.c
+++ b/mm/shmem.c
@@ -2345,7 +2345,7 @@ shmem_tmpfile(struct inode *dir, struct dentry *dentry, umode_t mode)
 			}
 		}
 #ifdef CONFIG_TMPFS_POSIX_ACL
-		error = generic_acl_init(inode, dir);
+		error = simple_acl_create(dir, inode);
 		if (error) {
 			iput(inode);
 			return error;
-- 
2.32.0

