diff --git a/drivers/video/msm/Kconfig b/drivers/video/msm/Kconfig
index a14862cac68..6cf4c9d176e 100644
--- a/drivers/video/msm/Kconfig
+++ b/drivers/video/msm/Kconfig
@@ -690,6 +690,9 @@ config FB_MSM_DEFAULT_DEPTH_ARGB8888
 config FB_MSM_DEFAULT_DEPTH_RGBA8888
 	bool "32 bits per pixel (RGBA8888)"
 
+config FB_MSM_DEFAULT_DEPTH_BGRA8888
+	bool "32 bits per pixel (BGRA8888)"
+
 endchoice
 
 config FB_MSM_EBI2_EPSON_S1D_QVGA_PANEL
diff --git a/drivers/video/msm/msm_fb.c b/drivers/video/msm/msm_fb.c
index a9cc8705345..444d7cd2beb 100644
--- a/drivers/video/msm/msm_fb.c
+++ b/drivers/video/msm/msm_fb.c
@@ -1553,6 +1553,25 @@ static int msm_fb_register(struct msm_fb_data_type *mfd)
 		bpp = 4;
 		break;
 
+	case MDP_BGRA_8888:
+		fix->type = FB_TYPE_PACKED_PIXELS;
+		fix->xpanstep = 1;
+		fix->ypanstep = 1;
+		var->vmode = FB_VMODE_NONINTERLACED;
+		var->blue.offset = 0;
+		var->green.offset = 8;
+		var->red.offset = 16;
+		var->blue.length = 8;
+		var->green.length = 8;
+		var->red.length = 8;
+		var->blue.msb_right = 0;
+		var->green.msb_right = 0;
+		var->red.msb_right = 0;
+		var->transp.offset = 24;
+		var->transp.length = 8;
+		bpp = 4;
+		break;
+
 	case MDP_RGBA_8888:
 		fix->type = FB_TYPE_PACKED_PIXELS;
 		fix->xpanstep = 1;
diff --git a/drivers/video/msm/msm_fb_def.h b/drivers/video/msm/msm_fb_def.h
index dcd648bab80..5623fec3caa 100644
--- a/drivers/video/msm/msm_fb_def.h
+++ b/drivers/video/msm/msm_fb_def.h
@@ -92,6 +92,8 @@ typedef unsigned int boolean;
 #define MSMFB_DEFAULT_TYPE MDP_ARGB_8888
 #elif defined(CONFIG_FB_MSM_DEFAULT_DEPTH_RGBA8888)
 #define MSMFB_DEFAULT_TYPE MDP_RGBA_8888
+#elif defined(CONFIG_FB_MSM_DEFAULT_DEPTH_BGRA8888)
+#define MSMFB_DEFAULT_TYPE MDP_BGRA_8888
 #else
 #define MSMFB_DEFAULT_TYPE MDP_RGB_565
 #endif
