From 3e3093e09a90ca804fce86b870cb02ed757dc40e Mon Sep 17 00:00:00 2001
From: Samuel Dionne-Riel <samuel@dionne-riel.com>
Date: Sat, 28 Mar 2020 02:02:16 -0400
Subject: [PATCH] HACK: Turn red LED on for pinephone

---
 board/sunxi/board.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/board/sunxi/board.c b/board/sunxi/board.c
index 6afea6ef42e..e5dd7cb7a9c 100644
--- a/board/sunxi/board.c
+++ b/board/sunxi/board.c
@@ -676,6 +676,8 @@ void sunxi_board_init(void)
 
 	sunxi_spl_store_dram_size(gd->ram_size);
 
+	gpio_direction_output(115, 1); // RED
+
 	/*
 	 * Only clock up the CPU to full speed if we are reasonably
 	 * assured it's being powered with suitable core voltage
-- 
2.23.1

