diff --git a/arch/arm/boot/dts/qcom/msm8953-potter.dtsi b/arch/arm/boot/dts/qcom/msm8953-potter.dtsi
index 9083a59a5ef7..2506e9e1d524 100644
--- a/arch/arm/boot/dts/qcom/msm8953-potter.dtsi
+++ b/arch/arm/boot/dts/qcom/msm8953-potter.dtsi
@@ -611,5 +611,5 @@
 		1 &intc 0 221 0
 		2 &tlmm 133 0>;
 	interrupt-names = "hc_irq", "pwr_irq", "status_irq";
-	cd-gpios = <&tlmm 133 0x0>;
+	cd-gpios = <&tlmm 133 0x1>;
 	qcom,cd-wakeup;

 	qcom,clk-rates = <400000 20000000 25000000 50000000 100000000
