diff --git a/arch/arm/boot/dts/qcom/Makefile b/arch/arm/boot/dts/qcom/Makefile
index c5a2e99104ab..3f421a0d8cd7 100644
--- a/arch/arm/boot/dts/qcom/Makefile
+++ b/arch/arm/boot/dts/qcom/Makefile
@@ -336,13 +336,9 @@ dtb-$(CONFIG_ARCH_MSM8953) += msm8953-addison-p1.dtb \
 endif
 
 ifeq ($(CONFIG_POTTER_DTB),y)
-dtb-$(CONFIG_ARCH_MSM8953) += msm8953-potter-p0a.dtb \
-	msm8953-potter-p1a.dtb \
-	msm8953-potter-p1b.dtb \
-	msm8953-potter-p2a.dtb \
-	msm8953-potter-p2a2.dtb \
-	msm8953-potter-p2b.dtb \
-	msm8953-potter-p3a.dtb
+dtb-$(CONFIG_ARCH_MSM8953) += msm8953-potter-p0a-p1a.dtb \
+	msm8953-potter-p1b-p2a.dtb \
+	msm8953-potter-p2a2-plus.dtb
 endif
 
 ifeq ($(CONFIG_ALBUS_DTB),y)
diff --git a/arch/arm/boot/dts/qcom/msm8953-potter-p0a-p1a.dts b/arch/arm/boot/dts/qcom/msm8953-potter-p0a-p1a.dts
new file mode 100644
index 000000000000..c64507c09e0a
--- /dev/null
+++ b/arch/arm/boot/dts/qcom/msm8953-potter-p0a-p1a.dts
@@ -0,0 +1,52 @@
+/* Copyright (c) 2015, Motorola Mobility Inc. All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 and
+ * only version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+/dts-v1/;
+
+
+#include "msm8953-potter.dtsi"
+
+
+/ {
+	qcom,board-id = <0x44 0x80a0>,<0x44 0x81a0>;
+};
+
+&tlmm {
+	usb_id_int_active {
+		usb_id_active: default {
+			mux {
+				pins = "gpio66";
+			};
+			config {
+				pins = "gpio66";
+			};
+		};
+	};
+	usb_id_int_sleep {
+		usb_id_sleep: sleep {
+			mux {
+				pins = "gpio66";
+			};
+			config {
+				pins = "gpio66";
+			};
+		};
+	};
+};
+&usb3 {
+	qcom,usbid-gpio = <&tlmm 66 0>;
+};
+
+&pm8953_diangu_dig {
+	qcom,cdc-boost-pdm-clk;
+};
+
diff --git a/arch/arm/boot/dts/qcom/msm8953-potter-p1b-p2a.dts b/arch/arm/boot/dts/qcom/msm8953-potter-p1b-p2a.dts
new file mode 100644
index 000000000000..e6f645c52e38
--- /dev/null
+++ b/arch/arm/boot/dts/qcom/msm8953-potter-p1b-p2a.dts
@@ -0,0 +1,26 @@
+/* Copyright (c) 2015, Motorola Mobility Inc. All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 and
+ * only version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+/dts-v1/;
+
+
+#include "msm8953-potter.dtsi"
+
+
+/ {
+	qcom,board-id = <0x44 0x81b0>,<0x44 0x82a0>,<0x45 0x82a0>;
+};
+
+&pm8953_diangu_dig {
+	qcom,cdc-boost-pdm-clk;
+};
+
diff --git a/arch/arm/boot/dts/qcom/msm8953-potter-p2a2-plus.dts b/arch/arm/boot/dts/qcom/msm8953-potter-p2a2-plus.dts
new file mode 100644
index 000000000000..fc1569b3a2ea
--- /dev/null
+++ b/arch/arm/boot/dts/qcom/msm8953-potter-p2a2-plus.dts
@@ -0,0 +1,24 @@
+/* Copyright (c) 2015, Motorola Mobility Inc. All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 and
+ * only version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+/dts-v1/;
+
+
+#include "msm8953-potter.dtsi"
+
+
+/ {
+	qcom,board-id = <0x45 0x82a2>,<0x45 0x82b0>,<0x46 0x83a0>,<0x46 0x83b0> ;
+};
+
+
+
