diff --git a/arch/arm/boot/dts/qcom/Makefile b/arch/arm/boot/dts/qcom/Makefile
index 6cdf596102e6..bda927287811 100644
--- a/arch/arm/boot/dts/qcom/Makefile
+++ b/arch/arm/boot/dts/qcom/Makefile
@@ -323,13 +323,9 @@ dtb-$(CONFIG_ARCH_MSM8953) += msm8953-addison-p1.dtb \
 endif
 
 ifeq ($(CONFIG_POTTER_DTB),y)
-dtb-$(CONFIG_ARCH_MSM8953) += msm8953-potter-p0a.dtb \
-	msm8953-potter-p1a.dtb \
-	msm8953-potter-p1b.dtb \
-	msm8953-potter-p2a.dtb \
-	msm8953-potter-p2a2.dtb \
-	msm8953-potter-p2b.dtb \
-	msm8953-potter-p3a.dtb
+dtb-$(CONFIG_ARCH_MSM8953) += msm8953-potter-p0a-p1a.dtb \
+	msm8953-potter-p1b-p2a.dtb
+	msm8953-potter-p2a2-plus.dtb
 endif
 
 ifeq ($(CONFIG_ALBUS_DTB),y)
