diff --git a/drivers/video/msm/mdss/mdss_fb.c b/drivers/video/msm/mdss/mdss_fb.c
index 6476e28970f6..23a8ec5131ac 100644
--- a/drivers/video/msm/mdss/mdss_fb.c
+++ b/drivers/video/msm/mdss/mdss_fb.c
@@ -3020,7 +3020,9 @@ static int mdss_fb_open(struct fb_info *info, int user)
 {
 	struct msm_fb_data_type *mfd = (struct msm_fb_data_type *)info->par;
 	struct mdss_fb_file_info *file_info = NULL;
+	struct fb_event event;
 	int result;
+	int blank;
 	struct task_struct *task = current->group_leader;
 
 	if (mfd->shutdown_pending) {
@@ -3047,13 +3049,22 @@ static int mdss_fb_open(struct fb_info *info, int user)
 	}
 
 	if (!mfd->ref_cnt) {
+		blank = FB_BLANK_UNBLANK;
+		event.info = info;
+		event.data = &blank;
+
+		fb_notifier_call_chain(FB_EARLY_EVENT_BLANK, &event);
+
 		result = mdss_fb_blank_sub(FB_BLANK_UNBLANK, info,
 					   mfd->op_enable);
 		if (result) {
+			fb_notifier_call_chain(FB_R_EARLY_EVENT_BLANK, &event);
 			pr_err("can't turn on fb%d! rc=%d\n", mfd->index,
 				result);
 			goto blank_error;
-		}
+		} else
+			fb_notifier_call_chain(FB_EVENT_BLANK, &event);
+
 	}
 
 	mfd->ref_cnt++;
