From e31df3ce072cab3b44d9074e57971babbeef8a45 Mon Sep 17 00:00:00 2001
From: Samuel Dionne-Riel <samuel@dionne-riel.com>
Date: Tue, 17 Dec 2019 00:03:24 -0500
Subject: [PATCH] Skip logger test with read-only source directory

---
 test/logger.rb | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/test/logger.rb b/test/logger.rb
index 259763a..5ec65f5 100644
--- a/test/logger.rb
+++ b/test/logger.rb
@@ -241,6 +241,8 @@ assert 'Logger.log', 'path' do
     logger.reopen
     assert_nothing_raised { logger << 'message' }
     assert_equal IO.read(file).split("\n")[1], 'messagemessage'
+  rescue => e
+    skip e.message
   ensure
     logger.close      if logger
     File.delete(file) if File.exist? file
-- 
2.23.0

