From a5ba367ef054aa7c24a778a4a535c34406e0be33 Mon Sep 17 00:00:00 2001
From: Samuel Dionne-Riel <samuel@dionne-riel.com>
Date: Wed, 9 Oct 2019 21:23:29 -0400
Subject: [PATCH] netfilter: xt_IDLETIMER: make compatible with USER_NS

---
 net/netfilter/xt_IDLETIMER.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/net/netfilter/xt_IDLETIMER.c b/net/netfilter/xt_IDLETIMER.c
index 94ce95e00cc..26f2279c391 100644
--- a/net/netfilter/xt_IDLETIMER.c
+++ b/net/netfilter/xt_IDLETIMER.c
@@ -365,7 +365,7 @@ static void reset_timer(const struct idletimer_tg_info *info,
 			read_lock_bh(&sk->sk_callback_lock);
 			if ((sk->sk_socket) && (sk->sk_socket->file) &&
 		    (sk->sk_socket->file->f_cred))
-				timer->uid = sk->sk_socket->file->f_cred->uid;
+				timer->uid = from_kuid_munged(current_user_ns(), sk->sk_socket->file->f_cred->uid);
 			read_unlock_bh(&sk->sk_callback_lock);
 		}
 
-- 
2.23.0

