diff --git a/pd-mapper.c b/pd-mapper.c
index 65c4b80..5e13608 100644
--- a/pd-mapper.c
+++ b/pd-mapper.c
@@ -193,11 +193,7 @@ static int pd_load_map(const char *file)
 	return 0;
 }
 
-#ifndef ANDROID
-#define FIRMWARE_BASE	"/lib/firmware/"
-#else
-#define FIRMWARE_BASE	"/vendor/firmware/"
-#endif
+#define FIRMWARE_BASE	"/run/current-system/sw/share/uncompressed-firmware/"
 
 static char *known_extensions[] = {
   ".jsn.xz",