diff --git a/translate.c b/translate.c
index de3baba..484db4e 100644
--- a/translate.c
+++ b/translate.c
@@ -22,11 +22,11 @@
 #define READWRITE_PATH	"/readwrite/"
 
 #ifndef ANDROID
-#define FIRMWARE_BASE	"/lib/firmware/"
+#define FIRMWARE_BASE	"/run/current-system/sw/share/uncompressed-firmware/"
 #define TQFTPSERV_TMP	"/tmp/tqftpserv"
 #define UPDATES_DIR	"updates/"
 #else
-#define FIRMWARE_BASE	"/vendor/firmware/"
+#define FIRMWARE_BASE	"/run/current-system/sw/share/uncompressed-firmware/"
 #define TQFTPSERV_TMP	"/data/vendor/tmp/tqftpserv"
 #endif
 