diff -rupbN linux.orig/arch/arm64/boot/dts/qcom/sm8250.dtsi linux/arch/arm64/boot/dts/qcom/sm8250.dtsi
--- linux.orig/arch/arm64/boot/dts/qcom/sm8250.dtsi	2024-10-19 13:35:28.332965210 +0000
+++ linux/arch/arm64/boot/dts/qcom/sm8250.dtsi	2024-10-19 17:04:33.824750269 +0000
@@ -608,7 +608,7 @@
 		};
 
 		cpu7_opp9: opp-1747200000 {
-			opp-hz = /bits/ 64 <1708800000>;
+			opp-hz = /bits/ 64 <1747200000>;
 			opp-peak-kBps = <5412000 42393600>;
 		};
 
