diff -rupbN linux.orig/arch/arm64/boot/dts/qcom/sm8250.dtsi linux/arch/arm64/boot/dts/qcom/sm8250.dtsi
--- linux.orig/arch/arm64/boot/dts/qcom/sm8250.dtsi	2024-10-19 13:35:28.332965210 +0000
+++ linux/arch/arm64/boot/dts/qcom/sm8250.dtsi	2024-10-26 01:57:44.836755804 +0000
@@ -2967,10 +2967,10 @@
 			gpu_opp_table: opp-table {
 				compatible = "operating-points-v2";

-				opp-670000000 {
-					opp-hz = /bits/ 64 <670000000>;
+				opp-650000000 {
+					opp-hz = /bits/ 64 <650000000>;
 					opp-level = <RPMH_REGULATOR_LEVEL_NOM_L1>;
-					opp-supported-hw = <0xa>;
+					opp-supported-hw = <0xb>;
 				};

 				opp-587000000 {
