diff --git a/mm/pagewalk.c b/mm/pagewalk.c
index 94c6272..12417f5 100644
--- a/mm/pagewalk.c
+++ b/mm/pagewalk.c
@@ -118,8 +118,12 @@
 	do {
 		next = hugetlb_entry_end(h, addr, end);
 		pte = huge_pte_offset(walk->mm, addr & hmask);
-		if (pte && walk->hugetlb_entry)
+
+		if (pte)
 			err = walk->hugetlb_entry(pte, hmask, addr, next, walk);
+		else if (walk->pte_hole)
+			err = walk->pte_hole(addr, next, walk);
+
 		if (err)
 			return err;
 	} while (addr = next, addr != end);
