From 504a81eeb1c6578988758cc321a5e699f714363c Mon Sep 17 00:00:00 2001
From: Samuel Dionne-Riel <samuel@dionne-riel.com>
Date: Wed, 24 Jun 2020 15:32:04 -0400
Subject: [PATCH 1/2] =?UTF-8?q?[HACK]=C2=A0Force=20teei=20paths=20into=20C?=
 =?UTF-8?q?FLAGS?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

---
 Makefile | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/Makefile b/Makefile
index d106a749c74a..3ee3d2d34903 100644
--- a/Makefile
+++ b/Makefile
@@ -1859,3 +1859,10 @@ FORCE:
 # Declare the contents of the .PHONY variable as phony.  We keep that
 # information in a variable so we can use it in if_changed and friends.
 .PHONY: $(PHONY)
+
+# What the heck mediatek...
+# The includes everywhere assume these folders are available. Mostly.
+# They are not.
+KBUILD_CFLAGS	+= -I$(srctree)/drivers/misc/mediatek/teei/300/common/include/
+KBUILD_CFLAGS	+= -I$(srctree)/drivers/misc/mediatek/teei/300/public/
+KBUILD_CFLAGS	+= -I$(srctree)/drivers/misc/mediatek/teei/300/tz_driver/include/
-- 
2.25.3


From f51ca10bc5409859ab0de1547c7db67ccfe64eb7 Mon Sep 17 00:00:00 2001
From: Samuel Dionne-Riel <samuel@dionne-riel.com>
Date: Wed, 24 Jun 2020 15:32:23 -0400
Subject: [PATCH 2/2] [HACK] teei header files are system headers

---
 drivers/misc/mediatek/cmdq/v3/cmdq_sec.c | 2 +-
 drivers/misc/mediatek/m4u/2.4/m4u.c      | 2 +-
 drivers/misc/mediatek/m4u/3.2/m4u.c      | 2 +-
 3 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/misc/mediatek/cmdq/v3/cmdq_sec.c b/drivers/misc/mediatek/cmdq/v3/cmdq_sec.c
index 96c394d4552c..c41f6d53bc57 100644
--- a/drivers/misc/mediatek/cmdq/v3/cmdq_sec.c
+++ b/drivers/misc/mediatek/cmdq/v3/cmdq_sec.c
@@ -80,7 +80,7 @@ static struct list_head gCmdqSecContextList;
 #include "cmdqsectl_api.h"
 
 #if defined(CONFIG_MICROTRUST_TEE_SUPPORT)
-#include "isee_kernel_api.h"
+#include <isee_kernel_api.h>
 #endif
 
 /* secure context to cmdqSecTL */
diff --git a/drivers/misc/mediatek/m4u/2.4/m4u.c b/drivers/misc/mediatek/m4u/2.4/m4u.c
index 770b6e5c8b06..805ee2aec64b 100644
--- a/drivers/misc/mediatek/m4u/2.4/m4u.c
+++ b/drivers/misc/mediatek/m4u/2.4/m4u.c
@@ -67,7 +67,7 @@
 
 
 #if defined(CONFIG_MICROTRUST_TEE_SUPPORT)
-#include "isee_kernel_api.h"
+#include <isee_kernel_api.h>
 #define DRM_M4U_DRV_DRIVER_ID   (0x977aa)
 #endif
 
diff --git a/drivers/misc/mediatek/m4u/3.2/m4u.c b/drivers/misc/mediatek/m4u/3.2/m4u.c
index c48cb65d0c87..349ed8e679f6 100644
--- a/drivers/misc/mediatek/m4u/3.2/m4u.c
+++ b/drivers/misc/mediatek/m4u/3.2/m4u.c
@@ -71,7 +71,7 @@
 
 
 #if defined(CONFIG_MICROTRUST_TEE_SUPPORT)
-#include "isee_kernel_api.h"
+#include <isee_kernel_api.h>
 #define DRM_M4U_DRV_DRIVER_ID   (0x977aa)
 #endif
 
-- 
2.25.3

